update operator-framework scenario for microservices (#144)
podtato-head/podtato-head: delivery/operator/README.md

# Deliver with an operator

Here's how to deliver podtato-head using the [Operator

> Framework](https://operatorframework.io) and a Helm chart-based operator built
> with its [Operator SDK](https://sdk.operatorframework.io). [Operator Lifecycle

Reviewer: ```suggestion
Framework](https://operatorframework.io) as a Helm chart-based operator built
with the [Operator SDK](https://sdk.operatorframework.io). [Operator Lifecycle
```

access token with `packages:write` permissions and running the following commands.

```bash
github_user=
github_token=

echo "${github_token}" | docker login --username=${github_user} --password-stdin ghcr.io

export IMAGE_TAG_BASE=ghcr.io/${github_user}/podtato-head/operator
export IMG=${IMAGE_TAG_BASE}:latest

make docker-build docker-push
make bundle bundle-build bundle-push
make catalog-build catalog-push
```

### Deploy the operator

To deploy the operator we'll deploy the catalog source and then subscribe to our
package, as follows. You may also want to review the test script at
[./test/deploy-operator.sh](./test/deploy-operator.sh).

First deploy the catalog source to the OLM namespace:

```bash
image_base_url=ghcr.io/${github_user}/podtato-head/operator
operator_version=0.0.1

kubectl apply -f - <<EOF
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
    name: podtato-head-catalog
    namespace: olm
spec:
    sourceType: grpc
    image: ${image_base_url}-catalog:v${operator_version}
    secrets:
      - ghcr
    displayName: podtato-head-catalog
    publisher: podtato-head
    updateStrategy:
        registryPoll:
            interval: 10m
EOF
```

> Next declare a subscription to the podtato-head-operator in the `operators` namespace:

Reviewer: I prefer creating a subscription as you've done here, but I believe that the SDK [provides a command](https://sdk.operatorframework.io/docs/olm-integration/quickstart-bundle/) to create install an operator which may be worth considering.

Author: @awgreene on review I think you may have been referring to the `make deploy` target [mentioned below](https://github.com/podtato-head/podtato-head/pull/144#discussion_r800913922). There's a section later in the doc which talks about that option, I updated that section's title to make it clearer.

Thanks!

```bash
operator_version=0.0.1

kubectl apply -f - <<EOF
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
    name: podtato-head-operator
    namespace: operators
spec:
    channel: alpha
    installPlanApproval: Automatic
    name: podtato-head-operator
    source: podtato-head-catalog
    sourceNamespace: olm
    startingCSV: podtato-head-operator.v${operator_version}
EOF
```

The subscription triggers creation of a ClusterServiceVersion (CSV) resource its
namespace, followed by an InstallPlan that deploys Deployments, Service

> Accounts, and other resources associated with the operator.

Reviewer: This is slightly off. Creating a Subscription causes the Catalog Operator to generate an installplan which records a list of resources that need to be installed in the `status.plan` array, the CSV is included in this list. The OLM operator will then reconcile the InstallPlan, creating each resource listed.

Watch for the successful provisioning of the operator deployment:

```bash

> kubectl get deployments --namespace operators --watch

Reviewer: Alternatively, users could watch for the podtato-head CSV to enter the `succeeded` phase.

Author: Added, thank you.

In an empty directory init a new Helm operator from the podtato-head chart as follows.

```bash
chart_dir=./delivery/chart
# expand relative path so it works from other dirs
chart_dir=$(cd ${chart_dir} && pwd)

work_dir=./work
mkdir -p ${work_dir}
cd ${work_dir}

operator-sdk init --plugins 'helm.sdk.operatorframework.io/v1' --project-version=3 \
    --project-name 'podtato-head-operator' \
    --domain=podtato-head.io \
    --group=apps \
    --version=v1alpha1 \
    --kind=PodtatoHeadApp \
    --helm-chart=${chart_dir}
```

Now build and push the operator image, the operator bundle, and a catalog which
includes the bundle as follows. We use GitHub's container registry `ghcr.io`;
you will need to be signed in to push images. Sign in by getting a personal
access token with `packages:write` permissions and running the following commands.

```bash
github_user=
github_token=

echo "${github_token}" | docker login --username=${github_user} --password-stdin ghcr.io

export IMAGE_TAG_BASE=ghcr.io/${github_user}/podtato-head/operator
export IMG=${IMAGE_TAG_BASE}:latest

make docker-build docker-push
make bundle bundle-build bundle-push
make catalog-build catalog-push
```

> ### Deploy the operator

Reviewer 2: By using operator-sdk to do that you could just tun the target `make deploy IMG=` or use operator-sdk run bundle <bundle-path> to check it with the OLM integration

Author: Thanks! I'll expand the README to mention that as a shortcut.

Author: @camilamacedo86 there's a section later in the doc which briefly walks through the make commands including `make deploy`. I updated that section's title to make it clearer. Thanks!
